Drivers/Src/Switches.c

```diff
@@ -19,125 +16,108 @@ static CPU_TS timestamp;
+    OSMutexCreate(&CommMutex, "communications Mutex", &err);
     assertOSError(0,err);
     BSP_SPI_Init();
     //Sets up pins 0-7 on GPIOA as input 
     uint8_t initTxBuf[3]={SPI_OPCODE_R, SPI_IODIRA, 0};
     uint8_t initRxBuf = 0;
-    GPIO_WriteBit(GPIOA, GPIO_Pin_4, Bit_RESET);
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, OFF);
     BSP_SPI_Write(initTxBuf,2);
     BSP_SPI_Read(&initRxBuf, 1);
-    GPIO_WriteBit(GPIOA, GPIO_Pin_4, Bit_SET);
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, ON);
     //OR Result of IODIRA read to set all to 1, then write it back to IODIRA
     initTxBuf[2] = initRxBuf|0xFF;
     initTxBuf[0] = SPI_OPCODE_W;
-    GPIO_WriteBit(GPIOA, GPIO_Pin_4, Bit_RESET);
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, OFF);
     BSP_SPI_Write(initTxBuf,3);
-    GPIO_WriteBit(GPIOA, GPIO_Pin_4, Bit_SET);
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, ON);
 
     //Sets up pin 7 on GPIOB as input (for ReverseSwitch)
     initTxBuf[0]=SPI_OPCODE_R;
     initTxBuf[1] = SPI_IODIRB;
     initTxBuf[2] = 0;
     initRxBuf = 0;
-    GPIO_WriteBit(GPIOA, GPIO_Pin_4, Bit_RESET);
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, OFF);
     BSP_SPI_Write(initTxBuf, 2);
     BSP_SPI_Read(&initRxBuf, 1);
-    GPIO_WriteBit(GPIOA, GPIO_Pin_4, Bit_SET);
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, ON);
     //OR IODIRB to set pin 7 to input and write it back
     initTxBuf[2] = initRxBuf|0x40;
     initTxBuf[0]=SPI_OPCODE_W;
-    GPIO_WriteBit(GPIOA, GPIO_Pin_4, Bit_RESET);
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, OFF);
     BSP_SPI_Write(initTxBuf,3);
-    GPIO_WriteBit(GPIOA, GPIO_Pin_4, Bit_SET);
-    OSMutexPost(
-        &SwitchMutex,
-        OS_OPT_POST_NONE,
-        &err
-    );
-    assertOSError(0,err);
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, ON);
 };
 
 /**
- * @brief   Writes to SPI with Write Bit set (no change allowed) to query GPIOA Register (x09), 
- * reads it back using SPI, isolates for input switch using AND mask and returns state 
+ * @brief   Reads from static variable bitmap holding values of switches
  * @param   sw
  * @return  ON/OFF State
  */ 
 State Switches_Read(switches_t sw){
-    uint8_t query[3]={SPI_OPCODE_R,SPI_GPIOA, 0}; //query GPIOA
-    uint8_t SwitchReadData = 0;
-    // If we are not trying to get the state of the ignition switches, or the Reverse switch
-    switch(sw) {
-    case CRUZ_ST:
-    case CRUZ_EN:
-    case REV_SW: 
-    case FOR_SW: 
-    case HEADLIGHT_SW: 
-    case LEFT_SW: 
-    case RIGHT_SW: 
-    case REGEN_SW:
+    return (State) ((SwitchStates_Bitmap >> sw) && 0x0001);
+}
+
+/**
+ * @brief   Sends SPI messages to read switches values. Also reads from GPIO's for 
+ *          ignition switch values
+ */ 
+void Switches_Update(void){
+    OS_ERR err;
+    CPU_TS timestamp;
+    uint8_t query[2]={SPI_OPCODE_R,SPI_GPIOA}; //query GPIOA
+    uint8_t SwitchDataReg1 = 0, SwitchDataReg2 = 0;
+
+    //Read all switches except for ignition and hazard
+    BSP_GPIO_Write_Pin(PORTA, GPIO_Pin_4, OFF);
```

Comment:
I assume this is the chip select pin. If so, this is part of the critical section and should be inside of the mutex